funsor/terms.py

```diff
@@ -1316,63 +1327,73 @@ class Independent(Funsor):
 
     This is equivalent to substitution followed by reduction::
 
-        f = ...
-        assert f.inputs['x'] == reals(4, 5)
+        f = ...  # a batched distribution
+        assert f.inputs['x_i'] == reals(4, 5)
         assert f.inputs['i'] == bint(3)
 
-        g = Independent(f, 'x', 'i')
+        g = Independent(f, 'x', 'i', 'x_i')
         assert g.inputs['x'] == reals(3, 4, 5)
+        assert 'x_i' not in g.inputs
         assert 'i' not in g.inputs
 
         x = Variable('x', reals(3, 4, 5))
-        g == f(x=x['i']).reduce(ops.logaddexp, 'i')
+        g == f(x_i=x['i']).reduce(ops.logaddexp, 'i')
     """
-    def __init__(self, fn, reals_var, bint_var):
+    def __init__(self, fn, reals_var, bint_var, diag_var):
         assert isinstance(fn, Funsor)
         assert isinstance(reals_var, str)
-        for k in fn.inputs:
-            if k == reals_var or k.startswith(reals_var + "__BOUND"):
-                reals_var_bound = k
-                break
-        assert reals_var_bound in fn.inputs
-        assert fn.inputs[reals_var_bound].dtype == 'real'
         assert isinstance(bint_var, str)
         assert bint_var in fn.inputs
         assert isinstance(fn.inputs[bint_var].dtype, int)
+        assert isinstance(diag_var, str)
+        assert diag_var in fn.inputs
+        assert fn.inputs[diag_var].dtype == 'real'
         inputs = fn.inputs.copy()
-        shape = (inputs.pop(bint_var).dtype,) + inputs.pop(reals_var_bound).shape
+        shape = (inputs.pop(bint_var).dtype,) + inputs.pop(diag_var).shape
+        assert reals_var not in inputs
```

Comment:
This doesn't forbid `reals_var == diag_var`, is that still allowed?

Reply:
Yes, that is still allowed.

Reply:
Can you add some test cases exercising that case?

Reply:
Done. Note this is pattern is used by `dist_to_funsor()` and is also tested in test/pyro/test_convert.py.